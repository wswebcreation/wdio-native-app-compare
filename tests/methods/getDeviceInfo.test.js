import { IMAGE_STRING } from '../mocks/mocks'

describe('getDeviceInfo', () => {
    let determineSmallIphoneSpy,
        determineLargeIphoneSpy,
        determineMediumIphoneSpy,
        determineExtraLargeIphoneSpy,
        getDeviceInfo,
        getScreenshotSizeSpy,
        Utils
    const screenData = {
        width: 10,
        height: 20,
    }

    beforeEach(() => {
        jest.isolateModules(() => {
            getDeviceInfo = require('../../lib/methods/getDeviceInfo').getDeviceInfo
            Utils = require('../../lib/helpers/utils')
        })
        delete global.driver
        global.driver = {
            capabilities: {},
            getWindowSize: jest.fn().mockResolvedValue(screenData),
        }
        getScreenshotSizeSpy = jest
            .spyOn(Utils, 'getScreenshotSize')
            .mockReturnValue({
                width: 1000,
                height: 2000,
            })
    })

    afterEach(() => {
        global.driver = {
            getWindowSize: jest.fn().mockRestore(),
        }
        determineSmallIphoneSpy.mockRestore()
        determineLargeIphoneSpy.mockRestore()
        determineMediumIphoneSpy.mockRestore()
        determineExtraLargeIphoneSpy.mockRestore()
        getScreenshotSizeSpy.mockRestore()
    })

    it('should be able to get the default device info for the initial run', async () => {
        determineSmallIphoneSpy = jest
            .spyOn(Utils, 'determineSmallIphone')
            .mockReturnValue(false)
        determineLargeIphoneSpy = jest
            .spyOn(Utils, 'determineLargeIphone')
            .mockReturnValue(false)
        determineMediumIphoneSpy = jest
            .spyOn(Utils, 'determineMediumIphone')
            .mockReturnValue(false)
        determineExtraLargeIphoneSpy = jest
            .spyOn(Utils, 'determineExtraLargeIphone')
            .mockReturnValue(false)

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
        expect(getScreenshotSizeSpy).toBeCalledWith(IMAGE_STRING)
        expect(global.driver.getWindowSize).toHaveBeenCalled()
        expect(determineSmallIphoneSpy).toBeCalledWith(screenData)
        expect(determineLargeIphoneSpy).toBeCalledWith(screenData)
        expect(determineMediumIphoneSpy).toBeCalledWith(screenData)
    })

    it('should be able to get the default device info for the second run', async () => {
        determineSmallIphoneSpy = jest
            .spyOn(Utils, 'determineSmallIphone')
            .mockReturnValue(false)
        determineLargeIphoneSpy = jest
            .spyOn(Utils, 'determineLargeIphone')
            .mockReturnValue(false)
        determineMediumIphoneSpy = jest
            .spyOn(Utils, 'determineMediumIphone')
            .mockReturnValue(false)
        determineExtraLargeIphoneSpy = jest
            .spyOn(Utils, 'determineExtraLargeIphone')
            .mockReturnValue(false)

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
        expect(getScreenshotSizeSpy).toBeCalledWith(IMAGE_STRING)
        expect(global.driver.getWindowSize).toHaveBeenCalled()
        expect(determineSmallIphoneSpy).toBeCalledWith(screenData)
        expect(determineLargeIphoneSpy).toBeCalledWith(screenData)
        expect(determineMediumIphoneSpy).toBeCalledWith(screenData)

        getScreenshotSizeSpy.mockRestore()
        determineSmallIphoneSpy.mockRestore()
        determineLargeIphoneSpy.mockRestore()
        determineMediumIphoneSpy.mockRestore()
        determineExtraLargeIphoneSpy.mockRestore()
        global.driver = {
            getWindowSize: jest.fn().mockRestore(),
        }

        // The second run to check that the data is stored in the `DEVICE_INFO` and all methods are not called again
        global.driver = {
            getWindowSize: jest.fn().mockResolvedValue(screenData),
        }
        getScreenshotSizeSpy = jest
            .spyOn(Utils, 'getScreenshotSize')
            .mockReturnValue({
                width: 1000,
                height: 2000,
            })
        determineSmallIphoneSpy = jest
            .spyOn(Utils, 'determineSmallIphone')
            .mockReturnValue(false)
        determineLargeIphoneSpy = jest
            .spyOn(Utils, 'determineLargeIphone')
            .mockReturnValue(false)
        determineMediumIphoneSpy = jest
            .spyOn(Utils, 'determineMediumIphone')
            .mockReturnValue(false)
        determineExtraLargeIphoneSpy = jest
            .spyOn(Utils, 'determineExtraLargeIphone')
            .mockReturnValue(false)

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
        expect(getScreenshotSizeSpy).not.toHaveBeenCalled()
        expect(global.driver.getWindowSize).not.toHaveBeenCalled()
        expect(determineSmallIphoneSpy).not.toHaveBeenCalled()
        expect(determineLargeIphoneSpy).not.toHaveBeenCalled()
        expect(determineMediumIphoneSpy).not.toHaveBeenCalled()
        expect(determineExtraLargeIphoneSpy).not.toHaveBeenCalled()
    })

    it('should be able to get the device info for an iPhone X', async () => {
        determineSmallIphoneSpy = jest
            .spyOn(Utils, 'determineSmallIphone')
            .mockReturnValue(true)
        determineLargeIphoneSpy = jest
            .spyOn(Utils, 'determineLargeIphone')
            .mockReturnValue(false)
        determineMediumIphoneSpy = jest
            .spyOn(Utils, 'determineMediumIphone')
            .mockReturnValue(false)
        determineExtraLargeIphoneSpy = jest
            .spyOn(Utils, 'determineExtraLargeIphone')
            .mockReturnValue(false)

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
    })

    it('should be able to get the device info for an iPhone 12', async () => {
        determineSmallIphoneSpy = jest
            .spyOn(Utils, 'determineSmallIphone')
            .mockReturnValue(false)
        determineLargeIphoneSpy = jest
            .spyOn(Utils, 'determineLargeIphone')
            .mockReturnValue(false)
        determineMediumIphoneSpy = jest
            .spyOn(Utils, 'determineMediumIphone')
            .mockReturnValue(true)
        determineExtraLargeIphoneSpy = jest
            .spyOn(Utils, 'determineExtraLargeIphone')
            .mockReturnValue(false)

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
    })

    it('should be able to get the device info for an iPhone X Large', async () => {
        determineSmallIphoneSpy = jest
            .spyOn(Utils, 'determineSmallIphone')
            .mockReturnValue(false)
        determineLargeIphoneSpy = jest
            .spyOn(Utils, 'determineLargeIphone')
            .mockReturnValue(true)
        determineMediumIphoneSpy = jest
            .spyOn(Utils, 'determineMediumIphone')
            .mockReturnValue(false)
        determineExtraLargeIphoneSpy = jest
            .spyOn(Utils, 'determineExtraLargeIphone')
            .mockReturnValue(false)

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
    })

    it('should be able to get the device info for an iPhone 12 Pro Max', async () => {
        determineSmallIphoneSpy = jest
            .spyOn(Utils, 'determineSmallIphone')
            .mockReturnValue(false)
        determineLargeIphoneSpy = jest
            .spyOn(Utils, 'determineLargeIphone')
            .mockReturnValue(false)
        determineMediumIphoneSpy = jest
            .spyOn(Utils, 'determineMediumIphone')
            .mockReturnValue(false)
        determineExtraLargeIphoneSpy = jest
            .spyOn(Utils, 'determineExtraLargeIphone')
            .mockReturnValue(true)

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
    })

    it('should be able to get the device info for a landscape phone', async () => {
        global.driver = {
            capabilities: {},
            getWindowSize: jest.fn().mockResolvedValue({
                width: 1000,
                height: 500,
            }),
        }

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
    })

    it('should be able to get the device info an Android phone', async () => {
        global.driver = {
            capabilities: {
                pixelRatio: 2.75,
                statBarHeight: 66,
                viewportRect: { left: 0, top: 66, width: 1000, height: 1802 },
            },
            getWindowSize: jest.fn().mockResolvedValue({
                width: 1000,
                height: 2000,
            }),
        }

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
    })
})
