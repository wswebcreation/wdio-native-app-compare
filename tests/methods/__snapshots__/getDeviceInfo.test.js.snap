// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDeviceInfo should be able to get the default device info for the initial run 1`] = `
{
  "dpr": 100,
  "isExtraLargeIphone": false,
  "isLargeIphone": false,
  "isMediumIphone": false,
  "isPortrait": true,
  "isSmallIphone": false,
  "rectangles": {
    "androidNavigationBar": {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    },
    "iOSHomeBar": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "statusBar": {
      "bottom": 2000,
      "left": 0,
      "right": 1000,
      "top": 0,
    },
  },
  "screenSize": {
    "height": 20,
    "width": 10,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;

exports[`getDeviceInfo should be able to get the default device info for the second run 1`] = `
{
  "dpr": 100,
  "isExtraLargeIphone": false,
  "isLargeIphone": false,
  "isMediumIphone": false,
  "isPortrait": true,
  "isSmallIphone": false,
  "rectangles": {
    "androidNavigationBar": {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    },
    "iOSHomeBar": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "statusBar": {
      "bottom": 2000,
      "left": 0,
      "right": 1000,
      "top": 0,
    },
  },
  "screenSize": {
    "height": 20,
    "width": 10,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;

exports[`getDeviceInfo should be able to get the default device info for the second run 2`] = `
{
  "dpr": 100,
  "isExtraLargeIphone": false,
  "isLargeIphone": false,
  "isMediumIphone": false,
  "isPortrait": true,
  "isSmallIphone": false,
  "rectangles": {
    "androidNavigationBar": {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    },
    "iOSHomeBar": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "statusBar": {
      "bottom": 2000,
      "left": 0,
      "right": 1000,
      "top": 0,
    },
  },
  "screenSize": {
    "height": 20,
    "width": 10,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;

exports[`getDeviceInfo should be able to get the device info an Android phone 1`] = `
{
  "dpr": 2.75,
  "isExtraLargeIphone": false,
  "isLargeIphone": false,
  "isMediumIphone": false,
  "isPortrait": true,
  "isSmallIphone": false,
  "rectangles": {
    "androidNavigationBar": {
      "bottom": 2000,
      "left": 0,
      "right": 1000,
      "top": 1868,
    },
    "iOSHomeBar": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "statusBar": {
      "bottom": 66,
      "left": 0,
      "right": 1000,
      "top": 0,
    },
  },
  "screenSize": {
    "height": 2000,
    "width": 1000,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;

exports[`getDeviceInfo should be able to get the device info for a landscape phone 1`] = `
{
  "dpr": 1,
  "isExtraLargeIphone": false,
  "isLargeIphone": false,
  "isMediumIphone": false,
  "isPortrait": false,
  "isSmallIphone": false,
  "rectangles": {
    "androidNavigationBar": {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    },
    "iOSHomeBar": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "statusBar": {
      "bottom": 20,
      "left": 0,
      "right": 1000,
      "top": 0,
    },
  },
  "screenSize": {
    "height": 500,
    "width": 1000,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;

exports[`getDeviceInfo should be able to get the device info for an iPhone 12 1`] = `
{
  "dpr": 100,
  "isExtraLargeIphone": false,
  "isLargeIphone": false,
  "isMediumIphone": true,
  "isPortrait": true,
  "isSmallIphone": false,
  "rectangles": {
    "androidNavigationBar": {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    },
    "iOSHomeBar": {
      "bottom": 7,
      "left": 125,
      "right": 140,
      "top": 830,
    },
    "statusBar": {
      "bottom": 4400,
      "left": 0,
      "right": 1000,
      "top": 0,
    },
  },
  "screenSize": {
    "height": 20,
    "width": 10,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;

exports[`getDeviceInfo should be able to get the device info for an iPhone 12 Pro Max 1`] = `
{
  "dpr": 100,
  "isExtraLargeIphone": true,
  "isLargeIphone": false,
  "isMediumIphone": false,
  "isPortrait": true,
  "isSmallIphone": false,
  "rectangles": {
    "androidNavigationBar": {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    },
    "iOSHomeBar": {
      "bottom": 7,
      "left": 136,
      "right": 155,
      "top": 912,
    },
    "statusBar": {
      "bottom": 4400,
      "left": 0,
      "right": 1000,
      "top": 0,
    },
  },
  "screenSize": {
    "height": 20,
    "width": 10,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;

exports[`getDeviceInfo should be able to get the device info for an iPhone X 1`] = `
{
  "dpr": 100,
  "isExtraLargeIphone": false,
  "isLargeIphone": false,
  "isMediumIphone": false,
  "isPortrait": true,
  "isSmallIphone": true,
  "rectangles": {
    "androidNavigationBar": {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    },
    "iOSHomeBar": {
      "bottom": 6,
      "left": 120,
      "right": 135,
      "top": 799,
    },
    "statusBar": {
      "bottom": 4400,
      "left": 0,
      "right": 1000,
      "top": 0,
    },
  },
  "screenSize": {
    "height": 20,
    "width": 10,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;

exports[`getDeviceInfo should be able to get the device info for an iPhone X Large 1`] = `
{
  "dpr": 100,
  "isExtraLargeIphone": false,
  "isLargeIphone": true,
  "isMediumIphone": false,
  "isPortrait": true,
  "isSmallIphone": false,
  "rectangles": {
    "androidNavigationBar": {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    },
    "iOSHomeBar": {
      "bottom": 6,
      "left": 133,
      "right": 148,
      "top": 883,
    },
    "statusBar": {
      "bottom": 4400,
      "left": 0,
      "right": 1000,
      "top": 0,
    },
  },
  "screenSize": {
    "height": 20,
    "width": 10,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;
